plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/form.tsx

```diff
@@ -72,11 +80,51 @@ const Form = < T extends AddressFormValues | ContactFormValues >( {
 		};
 	}, [ currentFields, currentFieldConfig, currentCountry, addressType ] );
 
+	const address2Field = addressFormFields.fields.find(
+		( field ) => field.key === 'address_2'
+	);
+
+	const hasAddress2FieldValue = useCallback( () => {
+		return objectHasProp( values, 'address_2' ) && values.address_2 !== '';
+	}, [ values ] );
+
+	const isAddress2FieldRequired = address2Field
+		? address2Field.required
+		: false;
+
+	const [ isAddress2FieldVisible, setIsAddress2FieldVisible ] = useState(
+		hasAddress2FieldValue || isAddress2FieldRequired
+	);
+
+	const toggleAddress2FieldVisibility = useCallback(
+		( event: React.MouseEvent< HTMLElement > ) => {
+			event.preventDefault();
+			setIsAddress2FieldVisible( ( prevVisibility ) => ! prevVisibility );
+		},
+		[]
+	);
+
+	const [ userChangedAddress2Field, setUserChangedAddress2Field ] =
+		useState( false );
+
 	// Stores refs for rendered fields so we can access them later.
 	const fieldsRef = useRef<
 		Record< string, ValidatedTextInputHandle | null >
 	>( {} );
 
+	// Toggle address 2 visibility.
+	useEffect( () => {
+		setIsAddress2FieldVisible(
+			isAddress2FieldRequired ||
+				hasAddress2FieldValue() ||
+				userChangedAddress2Field
+		);
+	}, [
+		isAddress2FieldRequired,
+		hasAddress2FieldValue,
+		userChangedAddress2Field,
+	] );
```

Comment:
This is really confusing to me, the triple condition as a start is confusing

Reply:
Thanks for calling this out as well, @senadir. This is the logic behind the triple condition:

- `isAddress2FieldRequired` → Did the merchant define this field as required in the page editor?
- `hasAddress2FieldValue` → Does this field have a value from a previous order?
- `userChangedAddress2Field` → Did the seller interact with this field during the current session?

What are your thoughts on introducing a constant before `setIsAddress2FieldVisible` and adding a comment for further explanation? I'm thinking about something like this:

```jsx
// Toggle address 2 visibility.
useEffect( () => {
	/**
	 * Determine if the address 2 input field should be shown based on the following conditions:
	 *
	 * 1. If the address 2 field is required.
	 * 2. If the address 2 field has a value.
	 * 3. If the seller has edited the address 2 field.
	 */
	const shouldShowAddress2Field =
		isAddress2FieldRequired ||
		hasAddress2FieldValue ||
		hasSellerChangedAddress2Field;

	setIsAddress2FieldVisible( shouldShowAddress2Field );
}, [
	isAddress2FieldRequired,
	hasAddress2FieldValue,
	hasSellerChangedAddress2Field,
] );
```

> [!NOTE]
> To increase the readability, I changed
> ```js
> const [ userChangedAddress2Field, setUserChangedAddress2Field ] = useState( false );
> ```
> to
> ```js
> const [ hasSellerChangedAddress2Field, setSellerChangedAddress2Field ] = useState( false );
> ```